src/rpm.rs

```diff
@@ -45,7 +54,71 @@ impl RPMPackage {
         out.write_all(&self.content)?;
         Ok(())
     }
+
+    /// sign all headers (except for the lead) using an external key and store it as the initial header
+    #[cfg(feature = "signing")]
+    pub fn sign(&mut self, secret_key: &[u8]) -> Result<(), RPMError> {
+        // create a temporary byte repr of the header
+        // and re-create all hashes
+        let mut header_bytes = Vec::<u8>::with_capacity(1024);
+        self.metadata.header.write(&mut header_bytes)?;
+
+        let _signature_size = header_bytes.len() + self.content.len();
+        let mut hasher = md5::Md5::default();
+
+        hasher.input(&header_bytes);
+        hasher.input(&self.content);
+
+        let hash_result = hasher.result();
+
+        let digest_md5 = hash_result.as_slice();
+
+        let digest_sha1 = sha1::Sha1::from(&header_bytes);
+        let digest_sha1 = digest_sha1.digest();
+
+        let signer = Signer::load_secret_key(secret_key)?;
+
+        // TODO verify the byte range is correct!
+        let signature_header = signer.sign(self.content.as_slice())?;
+        let _size = self.content.len();
+
+        let rsa_spanning_header_only = raw_signature_to_rfc4880(&signature_header[..])?;
+
+        let signature_header_and_archive = signer.sign(self.content.as_slice())?;
+        let size = self.content.len();
```

Comment:
This size is not correct. The size of this is defined as the header size plus the size of the compressed payload.

> RPMSIGTAG_SIZE | internal Header+Payload size in bytes.

(http://ftp.rpm.org/api/4.4.2.2/group__signature.html)
 So you need to add the size of the (non signature) header section as well

Reply:
this is the size calculation used before, just as an anchor:

```rust
        let mut header_bytes = Vec::new();
        header.write(&mut header_bytes)?;

        self.compressor = cpio::newc::trailer(self.compressor)?;
        let content = self.compressor.finish_compression()?;

        let signature_size = header_bytes.len() + content.len();
```

Reply:
an additional `let size = header_bytes.len() + size;` IIUC would fix it